### PR TITLE
Added ThirdPartyResource design doc and mockups

### PR DESCRIPTION
**Pull request by IanLewis**
This PR is a design document meant to address issue #1504.

### CODE REVIEW

**Comment by floreks** [comment]
Is anyone working on implementation of third party resources support in dashboard?

**Comment by bryk** [comment]
@IanLewis are you? @mijajasinski is going to help with reviewing

**Comment by IanLewis** [comment]
@floreks @bryk I haven't really started on the implementation if that's what you mean. I have a *very* rough mock on my local machine but that's it. I was starting to get wary of starting implementation since, while I know it's an important feature people are already using, TPR is still beta so the API may change and there is talk of adding [Aggregated API Servers (AAS)](https://github.com/kubernetes/dashboard/issues/1504#issuecomment-276543001) and I was thinking it may be best to hold off a bit and see how that plays out. WDYT?

I know we have added beta features like Deployments before, but TPR seems a bit more in the air to me. Maybe adding it but indicating it's still in beta might be best?

**Comment by IanLewis** [comment]
I also wanted to make sure we were cool good to merge this design doc PR before starting on the implementation in earnest.

**Comment by floreks** [comment]
@IanLewis I have worked a bit with API for TPR and first we need to migrate to client-go. Current k8s dependencies are a bit outdated and contain some old code for TPR. Without update we are able only to get basic info about TPR templates not objects created by them.

I have already started migration to newest client-go but it will take some time. Regarding AAS I'm not sure if they will help us with TPR. As stated it will be different way of extending API. More flexible and powerful. TPRs are easy to create and use but have some constraints.

File: docs/design/thirdpartyresources.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+Third Party Resources will be listed in the sidebar of the UI in a new section
+alongside first party objects. It will function similarly to existing menu
+items except that the items in the menu will be dynamically populated from the
+server.
+
+![](mockups/24-01-2017-thirdpartyresources/workloads.png)
+
+Clicking on a menu item brings up the Third Party Resource's detail page where
+the details of the ThirdPartyResource itself is displayed and objects created
```
Comment (mijajasinski) [inline docs/design/thirdpartyresources.md]: I think it should be "the details of the ThirdPartyResource itself are displayed"

**Comment by IanLewis** [comment]
@floreks 

> I have worked a bit with API for TPR and first we need to migrate to client-go. Current k8s dependencies are a bit outdated and contain some old code for TPR. Without update we are able only to get basic info about TPR templates not objects created by them.

Hmm. I've done some work using client-go as well, but I've not used the dashboard dependencies. I'll take your word that client-go is needed.

> I have already started migration to newest client-go but it will take some time. Regarding AAS I'm not sure if they will help us with TPR. As stated it will be different way of extending API. More flexible and powerful. TPRs are easy to create and use but have some constraints.

Yah, my understanding is that all of this is still quite a bit in flux and it's hard to see how it will land.

> Backend for this will be quite complex as it requires dynamic client creation for every TPR APIGroup. Similar to how it's done in kubectl.

Yah. It won't be super bad because the dynamic client will only be needed on a TPR detail page. TPRs themselves can be listed with the normal client so, if we go with the current design, the backend for the sidebar can just use the regular client.

**Comment by IanLewis** [comment]
For reference the issue for migrating to client-go is #1457

**Comment by IanLewis** [comment]
@floreks 

> Current k8s dependencies are a bit outdated and contain some old code for TPR. Without update we are able only to get basic info about TPR templates not objects created by them.

Can we just update the current dependencies for now to get the dynamic client? I'm afraid client-go will affect the implementation of other PRs like the terminal (#1345)

File: docs/design/thirdpartyresources.md (hunk 1 above)
Comment (IanLewis) [inline docs/design/thirdpartyresources.md]: Yah.

**Comment by floreks** [comment]
@IanLewis 

> Can we just update the current dependencies for now to get the dynamic client? I'm afraid client-go will affect the implementation of other PRs like the terminal (#1345)

I will check. Our deps are few months old and much has changed since then. It should be easier to update only k8s deps for sure.

**Comment by floreks** [comment]
@IanLewis @bryk Small update.

It appears that kubernetes is already using client-go as dependency and when using govendor to fetch latest deps we are also getting client-go and apimachinery. There is however [small issue on client-go](https://github.com/kubernetes/client-go/issues/93) side and it is not possible to build it right now. I wouldn't try to manually fix dependencies but rather wait until it gets fixed and we can safely use govendor to update and build them correctly.

Let's hope it will take few days not weeks to fix it :)

**Comment by IanLewis** [comment]
If you back out client-go to one prev revision it builds. I suspect they
updated client-go but not the apimachinery repo yet.

**Pull request merged**